src/Type/MixedType.php

```diff
@@ -462,8 +462,16 @@ function () use ($level): string {
 
 	public function toBoolean(): BooleanType
 	{
-		if ($this->subtractedType !== null && StaticTypeFactory::falsey()->equals($this->subtractedType)) {
-			return new ConstantBooleanType(true);
+		if ($this->subtractedType !== null) {
+			if (StaticTypeFactory::falsey()->equals($this->subtractedType)) {
+				return new ConstantBooleanType(true);
+			}
+			if ((new ConstantBooleanType(false))->isSuperTypeOf($this->subtractedType)->yes()) {
+				return new ConstantBooleanType(true);
+			}
+			if ((new ConstantBooleanType(true))->isSuperTypeOf($this->subtractedType)->yes()) {
+				return new ConstantBooleanType(false);
```

Comment:
I still feel this logic is wrong.

1) Let's say we subtract more than what's just falsy. Falsy is `0|0.0|''|'0'|array{}|false|null`. But if we subtract more than that, like `int|0.0|''|'0'|array{}|false|null` or `0|0.0|''|'0'|array{}|bool|null`, the result should still be the same.
2) `(new ConstantBooleanType(false))->isSuperTypeOf($this->subtractedType)->yes()` is also wrong. This represents `mixed~false`. But something from `mixed` could still be falsy.

Reply:
after more iterations, I think it turns out I am chasing ghosts and there is no bug.

the tests I came up with now don't fail any longer before this PR.
you may decide whether the tests are worth merging or we just close here, I guess?